modules/services/window-managers/i3-sway/sway.nix

```diff
@@ -0,0 +1,311 @@
+
+          Only basic keybinding is supported (bindsym keycomb action),
+          for more advanced setup use 'sway.extraConfig'.
+        '';
+      };
+    };
+  };
+
+  commonFunctions = import ./lib/functions.nix {
+    inherit cfg lib;
+    moduleName = "sway";
+  };
+
+  inherit (commonFunctions) keybindingsStr keycodebindingsStr modeStr assignStr barStr gapsStr floatingCriteriaStr windowCommandsStr colorSetStr;
+
+  startupEntryStr = { command, always, ... }: ''
+    ${if always then "exec_always" else "exec"} ${command}
+  '';
+
+  inputStr = name: attrs: ''
+    input "${name}" {
+    ${concatStringsSep "\n" (mapAttrsToList (name: value: "${name} ${value}") attrs)}
+    }
+  '';
+
+  outputStr = name: attrs: ''
+    output "${name}" {
+    ${concatStringsSep "\n" (mapAttrsToList (name: value: "${name} ${value}") attrs)}
+    }
+  '';
+
+  configFile = pkgs.writeText "sway.conf" ((if cfg.config != null then with cfg.config; ''
+    font pango:${concatStringsSep ", " fonts}
+    floating_modifier ${floating.modifier}
+    default_border ${if window.titlebar then "normal" else "pixel"} ${toString window.border}
+    default_floating_border ${if floating.titlebar then "normal" else "pixel"} ${toString floating.border}
+    hide_edge_borders ${window.hideEdgeBorders}
+    focus_wrapping ${if focus.forceWrapping then "yes" else "no"}
+    focus_follows_mouse ${if focus.followMouse then "yes" else "no"}
+    focus_on_window_activation ${focus.newWindow}
+    mouse_warping ${if focus.mouseWarping then "output" else "none"}
+    workspace_layout ${workspaceLayout}
+    workspace_auto_back_and_forth ${if workspaceAutoBackAndForth then "yes" else "no"}
+
+    client.focused ${colorSetStr colors.focused}
+    client.focused_inactive ${colorSetStr colors.focusedInactive}
+    client.unfocused ${colorSetStr colors.unfocused}
+    client.urgent ${colorSetStr colors.urgent}
+    client.placeholder ${colorSetStr colors.placeholder}
+    client.background ${colors.background}
+
+    ${keybindingsStr keybindings}
+    ${keycodebindingsStr keycodebindings}
+    ${concatStringsSep "\n" (mapAttrsToList inputStr input)}
+    ${concatStringsSep "\n" (mapAttrsToList outputStr output)}
+    ${concatStringsSep "\n" (mapAttrsToList modeStr modes)}
+    ${concatStringsSep "\n" (mapAttrsToList assignStr assigns)}
+    ${concatStringsSep "\n" (map barStr bars)}
+    ${optionalString (gaps != null) gapsStr}
+    ${concatStringsSep "\n" (map floatingCriteriaStr floating.criteria)}
+    ${concatStringsSep "\n" (map windowCommandsStr window.commands)}
+    ${concatStringsSep "\n" (map startupEntryStr startup)}
+  '' else "") + "\n" + (if cfg.systemdIntegration then ''
+    exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+  '' else "") + cfg.extraConfig);
+in
+
+{
+  options.wayland.windowManager.sway = {
+    enable = mkEnableOption "sway wayland compositor";
+
+    package = mkOption {
+      type = with types; nullOr package;
```

Comment:
Not sure why null would be a good fit here. What's the idea?

Reply:
I've installed sway using the nixos option which I believe configures some extra stuff, I didn't want this to be overwritten by the home-manager module

Reply:
This might no longer be necessary, I thought that the nixos sway module messed with some system thing that home-manager couldn't touch but maybe everything can be ported here